### PR TITLE
Solr 9 Update

### PR DESCRIPTION
**Pull request by JCRPaquin**
Welcome to Solr 9!

Until this PR Montysolr was running Solr 7.7.3. Solr-Lucene 9 introduced a number of improvements and optimizations that we benefit from indirectly, in addition to a number of changes that enable bug fixes.

This PR also enables us to use Java 17 to run Montysolr. Language version changes will likely come in a separate PR.

### CODE REVIEW

File: montysolr/src/test/java/org/apache/solr/search/TestAqpAdsabsSolrSearch.java [hunk 1]
```diff
@@ -1004,10 +992,10 @@ public void testSearch() throws Exception {
                 "spanNear([spanNear([all:acr::nasa, all:grant], 3, true), SpanMultiTermQueryWrapper(all:n*)], 5, false)",
                 SpanNearQuery.class);
         assertQueryEquals(req("defType", "aqp", "q", "\"NASA grant\"^0.9 NEAR N*"),
-                "spanNear([(spanNear([all:acr::nasa, all:grant], 1, true))^0.9, SpanMultiTermQueryWrapper(all:n*)], 5, false)",
+                "spanNear([PayloadScoreQuery(spanNear([all:acr::nasa, all:grant], 1, true), function: ConstantPayloadFunction, includeSpanScore: true), SpanMultiTermQueryWrapper(all:n*)], 5, false)",
```
Comment (kelockhart) [inline montysolr/src/test/java/org/apache/solr/search/TestAqpAdsabsSolrSearch.java:995]: Why did the `^0.9` get removed?

File: montysolr/src/test/java/org/apache/solr/search/TestAqpAdsabsSolrSearch.java [hunk 2]
```diff
@@ -1111,41 +1099,43 @@ public void testCustomScoring() throws Exception {
                         + "(keyword:\"dark energy\" | Synonym(keyword:acr::de keyword:syn::dark energy keyword:syn::de))",
                 BooleanQuery.class);
 
+        /*
```
Comment (kelockhart) [inline montysolr/src/test/java/org/apache/solr/search/TestAqpAdsabsSolrSearch.java:1102]: What's up with this test?

**Review by kelockhart — COMMENTED** [review]
No major concerns, mostly clarification questions for you

File: montysolr/src/main/java/org/apache/lucene/analysis/core/SelectiveLowerCaseFilterFactory.java [hunk 3]
```diff
@@ -18,9 +18,8 @@
  */
 
 import org.apache.lucene.analysis.TokenStream;
-import org.apache.lucene.analysis.util.AbstractAnalysisFactory;
-import org.apache.lucene.analysis.util.MultiTermAwareComponent;
```
[the remark this answers is not in the document]
Comment (JCRPaquin) [reply montysolr/src/main/java/org/apache/lucene/analysis/core/SelectiveLowerCaseFilterFactory.java:22]: IIRC it was removed in this version of Lucene and newer tools were added to replace it.

File: montysolr/src/main/java/org/apache/lucene/search/SecondOrderWeight.java [hunk 4]
```diff
@@ -31,11 +31,6 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
         return new SecondOrderListOfDocsScorer(innerWeight, hits, docBase);
     }
 
-    @Override
-    public void extractTerms(Set<Term> terms) {
```
[the remark this answers is not in the document]
Comment (JCRPaquin) [reply montysolr/src/main/java/org/apache/lucene/search/SecondOrderWeight.java:35]: IIRC this wasn't getting called anymore.

File: montysolr/src/test/java/org/apache/solr/search/TestAqpAdsabsSolrSearch.java [hunk 5]
```diff
@@ -585,9 +586,10 @@ public void testSpecialCases() throws Exception {
         assertQueryEquals(req("defType", "aqp", "q", "first_author:\"kurtz, m j\""),
                 "(first_author:kurtz, m j | first_author:kurtz, m j* | first_author:/kurtz, m[^ ]*/ | first_author:/kurtz, m[^ ]* j.*/ | first_author:kurtz, m | first_author:kurtz,)",
                 DisjunctionMaxQuery.class);
-        assertQueryEquals(req("defType", "aqp", "q", "author:\"^kurtz, m j\""),
-                "spanPosRange(spanOr([author:kurtz, m j, SpanMultiTermQueryWrapper(author:kurtz, m j*), SpanMultiTermQueryWrapper(author:/kurtz, m[^ ]*/), SpanMultiTermQueryWrapper(author:/kurtz, m[^ ]* j.*/), author:kurtz, m, author:kurtz,]), 0, 1)",
-                SpanPositionRangeQuery.class);
+        // TODO: Re-enable this after fixing SpanOr query string comparison
```
[the remark this answers is not in the document]
Comment (JCRPaquin) [reply montysolr/src/test/java/org/apache/solr/search/TestAqpAdsabsSolrSearch.java:589]: `spanOr` can be serialized in any order, which prevents direct string comparisons with the final query.

File: montysolr/src/test/java/org/apache/solr/search/TestAqpAdsabsSolrSearch.java [hunk 6]
```diff
@@ -705,67 +707,53 @@ public void testSpecialCases() throws Exception {
...
         assertQueryEquals(req("defType", "aqp", "q", "topn(5, (foo bar))"),
-                "custom(SecondOrderQuery(+all:foo +all:bar, collector=SecondOrderCollectorTopN(5)), sum(float(cite_read_boost),const(0.5)))",
-                CustomScoreQuery.class);
+                "FunctionScoreQuery(SecondOrderQuery(+all:foo +all:bar, collector=SecondOrderCollectorTopN(5)), scored by boost(sum(float(cite_read_boost),const(0.5))))",
+                FunctionScoreQuery.class);
 
-        assertQueryEquals(req("defType", "aqp", "q", "topn(5, edismax(dog OR cat))", "qf", "title^1 abstract^0.5"),
-                "custom(SecondOrderQuery(((abstract:dog)^0.5 | title:dog) ((abstract:cat)^0.5 | title:cat), collector=SecondOrderCollectorTopN(5)), sum(float(cite_read_boost),const(0.5)))",
-                CustomScoreQuery.class);
+//        assertQueryEquals(req("defType", "aqp", "q", "topn(5, edismax(dog OR cat))", "qf", "title^1 abstract^0.5"),
```
[the remark this answers is not in the document]
Comment (JCRPaquin) [reply montysolr/src/test/java/org/apache/solr/search/TestAqpAdsabsSolrSearch.java:751]: `spanOr` string comparison issue

File: montysolr/src/test/java/org/apache/solr/search/TestAqpAdsabsSolrSearch.java (hunk 1 above)
Comment (JCRPaquin) [reply montysolr/src/test/java/org/apache/solr/search/TestAqpAdsabsSolrSearch.java:995]: It should be incorporated into the `PayloadScoreQuery`, which modifies the score. Unfortunately, the constant scoring function is opaque and doesn't serialize its value (0.9 in this case).

File: montysolr/src/test/java/org/apache/solr/search/TestAqpAdsabsSolrSearch.java (hunk 2 above)
Comment (JCRPaquin) [reply montysolr/src/test/java/org/apache/solr/search/TestAqpAdsabsSolrSearch.java:1102]: Similar issue to `spanOr`: terms can be in any order when the query is serialized, preventing naive, exact string comparison.

**Pull request merged**